llpc/lower/llpcSpirvLowerTerminator.cpp

```diff
@@ -117,25 +117,26 @@ void SpirvLowerTerminator::visitCallInst(CallInst &callInst) {
   if (isa<ReturnInst>(*instIter))
     return;
 
-  // If terminator branches to block containing a PHI node which references
-  // this block then instructions following kill may be part of a return
-  // value sequence. In which case we cannot safely modify the instructions.
-  auto blockTerm = parentBlock->getTerminator();
-  if (!blockTerm)
-    return;
-  if (BranchInst *branch = dyn_cast<BranchInst>(blockTerm)) {
-    assert(branch->isUnconditional());
-    BasicBlock *targetBlock = branch->getSuccessor(0);
-    for (PHINode &phiNode : targetBlock->phis()) {
-      for (BasicBlock *incomingBlock : phiNode.blocks()) {
-        if (incomingBlock == parentBlock)
-          return;
+  // We must update any phis that point to this block as stale values confuse later passes
+  for (auto succBlock : successors(parentBlock)) {
+    for (auto &phiNode : succBlock->phis()) {
+      phiNode.removeIncomingValue(parentBlock, false);
+      if (phiNode.getNumOperands() == 0) {
+        // PHI began empty which means this successor block is now unreachable.
+        // Update all references to the PHI with poison and mark it for removal.
+        // We can do this because all references to this PHI are now also unreachable,
+        // or are PHI nodes which will be cleaned when the CFG is simplified.
+        auto poisonValue = PoisonValue::get(phiNode.getType());
+        phiNode.replaceAllUsesWith(poisonValue);
+        m_removalStack.emplace_back(&phiNode);
       }
     }
   }
 
-  // Add return
-  ReturnInst::Create(*m_context, nullptr, &*instIter);
+  // Add return (of poison)
+  auto returnType = parentBlock->getParent()->getReturnType();
+  auto returnValue = PoisonValue::get(returnType);
+  ReturnInst::Create(*m_context, returnValue, &*instIter);
```

Comment:
Should this be conditional depending on whether the return type is void or not?

Reply:
That's true.  Although this was working apparently.
One oddity I have noticed in LLVM recently is that `void poison` is a valid type to create in the IR data structures, but is illegal to parse.  I suspect this might be a case where this works because `void` is actually a type in the type system, so `getReturnType()` is not `nullptr`.
Anyway, I will fix it.